File: 019. Remove Nth Node From End of List/Python/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        first = head
        second = head
        prev = head
        while n:
            first = first.next
            n -= 1
        while first:
            first = first.next
            prev = second
            second = second.next
        if second == head:
            return head.next
        prev.next = prev.next.next
        return head

File: 019. Remove Nth Node From End of List/Java/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode first = head;
        ListNode second = head;
        ListNode prev = head;
        while (n != 0) {
            first = first.next;
            n--;
        }
        while (first != null) {
            first = first.next;
            prev = second;
            second = second.next;
        }
        if (second == head) {
            return head.next;
        }
        prev.next = prev.next.next;
        return head;
    }
}